KAFKA-16260: Deprecate window.size.ms and window.inner.class.serde in StreamsConfig

FrankYang0529 opened the pull request:
The `window.size.ms` and `window.inner.class.serde` are not a true KafkaStreams config, and are ignored when set from a KStreams application. Both belong on the client.

KIP-1020: https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=290982804
Reviewers: Lucas Brutschy <lbrutschy@confluent.io>

FrankYang0529 committed:
KAFKA-16260: Deprecate window.size.ms and window.inner.class.serde in StreamsConfig

Signed-off-by: PoAn Yang <payang@apache.org>

FrankYang0529 committed:
Merge branch 'trunk' into KAFKA-16260

lucasbru commented on docs/upgrade.html:
```diff
@@ -430,6 +430,9 @@ <h5><a id="upgrade_servers_400_notable" href="#upgrade_servers_400_notable">Nota
                     <a href="https://cwiki.apache.org/confluence/x/8ItyEg">KIP-1050</a> for more details </li>
                 </ul>
+                <li>
+                    The <code>window.size.ms</code> and <code>window.inner.serde.class</code> in stream config are deprecated.
```
This should not be in the 4.0 section.

lucasbru replied:
```suggestion
                    The <code>window.size.ms</code> and <code>window.inner.serde.class</code> in `StreamsConfig` are deprecated. Use the corresponding string constants defined in `TimeWindowedSerializer`, `TimeWindowedDeserializer`, `SessionWindowedSerializer` and `SessionWindowedDeserializer` instead.
```

Something like that? Could also be the actual string constants. But it's good to make these notes actionable.

lucasbru commented on streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:
```diff
@@ -823,13 +827,28 @@ public class StreamsConfig extends AbstractConfig {
+    /**
+     * {@code windowed.inner.class.serde}
+     *
+     * @deprecated since 4.0.0.
```
4.1.0

lucasbru commented on streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:
```diff
@@ -823,13 +827,28 @@ public class StreamsConfig extends AbstractConfig {
+    /**
+     * {@code window.size.ms}
+     *
+     * @deprecated since 4.0.0.
```
4.1.0

lucasbru commented on streams/src/main/java/org/apache/kafka/streams/kstream/TimeWindowedSerializer.java:
```diff
@@ -38,32 +48,42 @@ public TimeWindowedSerializer(final Serializer<T> inner) {
-        Serde<T> windowInnerClassSerde = null;
-        if (windowedInnerClassSerdeConfig != null) {
+        String serializerConfigFrom = WINDOWED_INNER_SERIALIZER_CLASS;
+        String windowedInnerSerializerClassConfig = (String) configs.get(WINDOWED_INNER_SERIALIZER_CLASS);
```
variable name suggestion: serializerConfigValue

Also in other places. I think the key/value relationship between these two variables is helpful to understand this code.

lucasbru commented on streams/src/main/java/org/apache/kafka/streams/kstream/TimeWindowedSerializer.java:
```diff
@@ -38,32 +48,42 @@ public TimeWindowedSerializer(final Serializer<T> inner) {
-        final String windowedInnerClassSerdeConfig = (String) configs.get(StreamsConfig.WINDOWED_INNER_CLASS_SERDE);
-        Serde<T> windowInnerClassSerde = null;
-        if (windowedInnerClassSerdeConfig != null) {
+        String serializerConfigFrom = WINDOWED_INNER_SERIALIZER_CLASS;
```
variable name suggestion: serializerConfigKey

Also in other places

lucasbru commented on streams/src/main/java/org/apache/kafka/streams/kstream/SessionWindowedDeserializer.java:
```diff
@@ -23,10 +23,20 @@
 public class SessionWindowedDeserializer<T> implements Deserializer<Windowed<T>> {
 
+    /**
+     * Configuration key for the windowed inner deserializer class.
```
Can we put any more information into this (and the other) java doc comments about the constants? Since this is a public string constant now, the use of the constant should ideally be clear from the javadoc comments.

lucasbru reviewed (commented):
Thanks for the PR, @FrankYang0529 ! Overall this looks like a well-done PR to me. I left some comments.

FrankYang0529 committed:
Merge branch 'trunk' into KAFKA-16260

FrankYang0529 committed:
address comment

Signed-off-by: PoAn Yang <payang@apache.org>

FrankYang0529 commented:
@lucasbru Thanks for the review. I addressed all comments. The document update is like following:

<img width="555" alt="Screenshot 2025-03-26 at 12 50 26 PM" src="https://github.com/user-attachments/assets/c79c3677-a7a1-423b-a13b-deb3fda87e84" />

lucasbru reviewed (approved):
LGTM, thanks!

mjsax commented:
FYI: https://github.com/apache/kafka/pull/19570

I usually prefer to verify the error message in tests, to avoid incorrect setups...

This would be better
```
final IllegalArgumentException expected = assertThrows(IllegalArgumentException.class, () -> serializer.configure(props, false));
assertThat(e.getMessage(), is("..."));
```

Did not want to go the extra mile and add it everything after the fact, but only fixed the issue itself.

FrankYang0529 commented:
@mjsax Thanks for the follow-up PR. I will pay more attention to the warning messages next time. Do you want to add verification about error message in #19570? Or I can create another PR for it. Thanks.

mjsax commented:
No worries. All good. -- Just sharing to general awareness. Not meant as critique.

> Do you want to add verification about error message in https://github.com/apache/kafka/pull/19570?

No, I did not plan to do this.

> Or I can create another PR for it.

If you want, feel free to do a follow up PR after my PR is merged. But know that we did fix it, it might also be ok to let it be as-is. Up to you. If you feel like it, and want to make this improvement, be my guest. Might actually apply to many other tests, too.
